### PR TITLE
Rc 3.1.2 patch

### PR DESCRIPTION
**Pull request by ozada**

### CODE REVIEW

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Clones added
============
- view/frontend/web/js/view/payment/method-renderer/vault.js  2
- view/frontend/web/js/view/payment/method-renderer/creditcard.js  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175&bid=17329626)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 2
           

Complexity increasing per file
==============================
- view/frontend/web/js/view/payment/seamless-vault-enabler.js  1
- view/frontend/web/js/view/payment/method-renderer/constants.js  1
- view/frontend/web/js/view/payment/method-renderer/seamlessformutils.js  4
- Block/Checkout/CreditCardThreeDForm.php  1
         

Complexity decreasing per file
==============================
+ view/frontend/web/js/view/payment/method-renderer/creditcard.js  -2
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -1
         

Clones removed
==============
+ view/frontend/web/js/view/payment/method-renderer/creditcard.js  -6
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175&bid=17329626)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 2
           

Complexity increasing per file
==============================
- view/frontend/web/js/view/payment/seamless-vault-enabler.js  1
- view/frontend/web/js/view/payment/method-renderer/seamlessformutils.js  4
- view/frontend/web/js/view/payment/method-renderer/constants.js  1
- Block/Checkout/CreditCardThreeDForm.php  1
         

Complexity decreasing per file
==============================
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -1
+ view/frontend/web/js/view/payment/method-renderer/creditcard.js  -2
         

Clones removed
==============
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -6
+ view/frontend/web/js/view/payment/method-renderer/creditcard.js  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175&bid=17329626)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 3
           

Complexity increasing per file
==============================
- Block/Checkout/CreditCardThreeDForm.php  1
- view/frontend/web/js/view/payment/seamless-vault-enabler.js  1
- view/frontend/web/js/view/payment/method-renderer/constants.js  1
- view/frontend/web/js/view/payment/method-renderer/seamlessformutils.js  4
         

Complexity decreasing per file
==============================
+ view/frontend/web/js/view/payment/method-renderer/creditcard.js  -2
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -1
         

Clones removed
==============
+ view/frontend/web/js/view/payment/method-renderer/creditcard.js  -6
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175&bid=17329626)

File: Gateway/Request/AccountInfoFactory.php [hunk 1]
```diff
@@ -123,19 +125,22 @@ private function setUserCreationData($accountInfo)
...
             $createdDates = $this->vaultCollection->addFieldToFilter('gateway_token', $token)
                 ->getColumnValues(self::CREATED_AT_KEY);
-
-            $creationDate = new \DateTime(reset($createdDates));
-            $creationDate->format(AccountInfo::DATE_FORMAT);
-
-            $accountInfo->setCardCreationDate($creationDate);
+            $creationDate->modify(reset($createdDates));
         }
+        $date = \DateTime::createFromFormat(
```
Comment (rinnhofer) [inline Gateway/Request/AccountInfoFactory.php:139]: Codacy Issue found: [Avoid using static access to class '\DateTime' in method 'setCreditCardCreationDate'.](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 3
           

Coverage increased per file
===========================
+ Gateway/Request/BasketFactory.php  1
         

Coverage decreased per file
===========================
- Controller/Frontend/Creditcard.php  -2
         

Complexity increasing per file
==============================
- Gateway/Helper/AccountHolderMapper.php  4
- view/frontend/web/js/view/payment/method-renderer/constants.js  1
- view/frontend/web/js/view/payment/method-renderer/creditcard.js  2
- Block/Checkout/CreditCardThreeDForm.php  1
- view/frontend/web/js/view/payment/seamless-vault-enabler.js  1
- view/frontend/web/js/view/payment/method-renderer/seamlessformutils.js  4
         

Complexity decreasing per file
==============================
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -1
         

Clones removed
==============
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -6
+ view/frontend/web/js/view/payment/method-renderer/creditcard.js  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175&bid=17329626)

File: view/frontend/web/js/view/payment/method-renderer/seamlessformutils.js [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+            window.scrollTo(0,0);
+            disableButtonById.call(this);
+            let responseKeys = Object.keys(response);
+            let hasMessages = false;
+            responseKeys.forEach(
+                function ( responseKey ) {
+                    if (responseKey.startsWith(SeamlessCreditCardConstants.wpp.errorPrefix)) {
+                        hasMessages = true;
+                        messageList.addErrorMessage({
+                            message: response[responseKey]
```
Comment (rinnhofer) [inline view/frontend/web/js/view/payment/method-renderer/seamlessformutils.js:126]: Codacy Issue found: [Generic Object Injection Sink](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175)

File: view/frontend/web/js/view/payment/seamless-vault-enabler.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+        "Magento_Vault/js/view/payment/vault-enabler",
+    ],
+    function (VaultEnabler) {
+        return VaultEnabler.extend({
+            defaults: {
+                isActivePaymentTokenEnabler: false
+            }
+        });
+    }
+)
```
Comment (rinnhofer) [inline view/frontend/web/js/view/payment/seamless-vault-enabler.js:18]: Codacy Issue found: [Missing semicolon.](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 3
           

Complexity increasing per file
==============================
- Block/Checkout/CreditCardThreeDForm.php  1
- view/frontend/web/js/view/payment/method-renderer/constants.js  1
- view/frontend/web/js/view/payment/method-renderer/seamlessformutils.js  4
- Gateway/Helper/AccountHolderMapper.php  4
- view/frontend/web/js/view/payment/method-renderer/creditcard.js  2
- view/frontend/web/js/view/payment/seamless-vault-enabler.js  1
         

Complexity decreasing per file
==============================
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -1
         

Clones removed
==============
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -6
+ view/frontend/web/js/view/payment/method-renderer/creditcard.js  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175&bid=17329626)

File: view/frontend/web/js/view/payment/method-renderer/seamlessformutils.js (hunk 2 above)
Comment (rinnhofer) [inline view/frontend/web/js/view/payment/method-renderer/seamlessformutils.js:126]: Codacy Issue found: [Generic Object Injection Sink](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175)

File: view/frontend/web/js/view/payment/seamless-vault-enabler.js (hunk 3 above)
Comment (rinnhofer) [inline view/frontend/web/js/view/payment/seamless-vault-enabler.js:18]: Codacy Issue found: [Missing semicolon.](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175)

File: Gateway/Request/AccountInfoFactory.php (hunk 1 above)
Comment (rinnhofer) [inline Gateway/Request/AccountInfoFactory.php:139]: Codacy Issue found: [Avoid using static access to class '\DateTime' in method 'setCreditCardCreationDate'.](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 3
           

Coverage increased per file
===========================
+ Gateway/Request/BasketFactory.php  1
         

Coverage decreased per file
===========================
- Controller/Frontend/Creditcard.php  -2
         

Complexity increasing per file
==============================
- view/frontend/web/js/view/payment/method-renderer/seamlessformutils.js  4
- view/frontend/web/js/view/payment/method-renderer/creditcard.js  2
- Gateway/Helper/AccountHolderMapper.php  4
- Block/Checkout/CreditCardThreeDForm.php  1
- view/frontend/web/js/view/payment/method-renderer/constants.js  1
- view/frontend/web/js/view/payment/seamless-vault-enabler.js  1
         

Complexity decreasing per file
==============================
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -1
         

Clones removed
==============
+ view/frontend/web/js/view/payment/method-renderer/creditcard.js  -6
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175&bid=17329626)

File: Gateway/Request/AccountInfoFactory.php (hunk 1 above)
Comment (rinnhofer) [inline Gateway/Request/AccountInfoFactory.php:139]: Codacy Issue found: [Avoid using static access to class '\DateTime' in method 'setCreditCardCreationDate'.](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175)

File: view/frontend/web/js/view/payment/method-renderer/seamlessformutils.js (hunk 2 above)
Comment (rinnhofer) [inline view/frontend/web/js/view/payment/method-renderer/seamlessformutils.js:126]: Codacy Issue found: [Generic Object Injection Sink](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175)

File: view/frontend/web/js/view/payment/seamless-vault-enabler.js (hunk 3 above)
Comment (rinnhofer) [inline view/frontend/web/js/view/payment/seamless-vault-enabler.js:18]: Codacy Issue found: [Missing semicolon.](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175)

**Comment by tatsta** [comment]
I have no idea why Travis run on PR is failing - but triggering a manual run (see Travis Branch run) works ok. So I suggest not to spend time on this topic, since there is a change in UI test configuration coming up. If this is the only obstacle to merge this in, go ahead and merge it :)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 3
           

Coverage increased per file
===========================
+ Gateway/Request/BasketFactory.php  1
         

Coverage decreased per file
===========================
- Controller/Frontend/Creditcard.php  -2
         

Complexity increasing per file
==============================
- Gateway/Helper/AccountHolderMapper.php  4
- view/frontend/web/js/view/payment/seamless-vault-enabler.js  1
- view/frontend/web/js/view/payment/method-renderer/creditcard.js  2
- Block/Checkout/CreditCardThreeDForm.php  1
- view/frontend/web/js/view/payment/method-renderer/constants.js  1
- view/frontend/web/js/view/payment/method-renderer/seamlessformutils.js  4
         

Complexity decreasing per file
==============================
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -1
         

Clones removed
==============
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -6
+ view/frontend/web/js/view/payment/method-renderer/creditcard.js  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175&bid=17329626)

File: view/frontend/web/js/view/payment/method-renderer/seamlessformutils.js (hunk 2 above)
Comment (rinnhofer) [inline view/frontend/web/js/view/payment/method-renderer/seamlessformutils.js:126]: Codacy Issue found: [Generic Object Injection Sink](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175)

File: view/frontend/web/js/view/payment/seamless-vault-enabler.js (hunk 3 above)
Comment (rinnhofer) [inline view/frontend/web/js/view/payment/seamless-vault-enabler.js:18]: Codacy Issue found: [Missing semicolon.](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175)

File: Gateway/Request/AccountInfoFactory.php (hunk 1 above)
Comment (rinnhofer) [inline Gateway/Request/AccountInfoFactory.php:139]: Codacy Issue found: [Avoid using static access to class '\DateTime' in method 'setCreditCardCreationDate'.](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
           

Coverage increased per file
===========================
+ Gateway/Request/BasketFactory.php  1
         

Coverage decreased per file
===========================
- Controller/Frontend/Creditcard.php  -2
         

Complexity increasing per file
==============================
- view/frontend/web/js/view/payment/method-renderer/seamlessformutils.js  4
- Gateway/Helper/AccountHolderMapper.php  4
- Block/Checkout/CreditCardThreeDForm.php  1
- view/frontend/web/js/view/payment/seamless-vault-enabler.js  1
- view/frontend/web/js/view/payment/method-renderer/constants.js  1
- view/frontend/web/js/view/payment/method-renderer/creditcard.js  2
         

Complexity decreasing per file
==============================
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -1
         

Clones removed
==============
+ view/frontend/web/js/view/payment/method-renderer/creditcard.js  -6
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175&bid=17329626)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
           

Coverage increased per file
===========================
+ Gateway/Request/BasketFactory.php  1
         

Coverage decreased per file
===========================
- Controller/Frontend/Creditcard.php  -2
         

Complexity increasing per file
==============================
- view/frontend/web/js/view/payment/seamless-vault-enabler.js  1
- view/frontend/web/js/view/payment/method-renderer/creditcard.js  2
- Block/Checkout/CreditCardThreeDForm.php  1
- view/frontend/web/js/view/payment/method-renderer/constants.js  1
- view/frontend/web/js/view/payment/method-renderer/seamlessformutils.js  4
- Gateway/Helper/AccountHolderMapper.php  4
         

Complexity decreasing per file
==============================
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -1
         

Clones removed
==============
+ view/frontend/web/js/view/payment/method-renderer/creditcard.js  -6
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175&bid=17329626)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
           

Coverage decreased per file
===========================
- Gateway/Helper/Order.php  -100
- Controller/Adminhtml/Test/Credentials.php  -100
- Controller/Frontend/Notify.php  -100
- Gateway/Config/CanVoidHandler.php  -100
- Model/Adminhtml/Support.php  -99
- Gateway/Helper/CalculationTrait.php  -83
- Gateway/Request/CreditCardTransactionFactory.php  -55
- Controller/Adminhtml/Support/Contact.php  -100
- Gateway/Request/AccountHolderFactory.php  -81
- Block/Checkout/SepaMandateBlock.php  -100
- Block/Adminhtml/Support/Contact.php  -100
- Gateway/Validator/AddressValidatorFactory.php  -100
- Gateway/Validator/ValidatorFactory.php  -100
- Gateway/Model/Notify.php  -90
- Gateway/Request/PayByBankAppTransactionFactory.php  -98
- Model/Ui/TokenUiComponentProvider.php  -95
- Gateway/Helper/NestedObject.php  -100
- Gateway/Request/GiropayTransactionFactory.php  -100
- Gateway/Request/PoiPiaTransactionFactory.php  -100
- Controller/Frontend/Creditcard.php  -75
- Gateway/Request/TransactionFactory.php  -89
- Gateway/Model/TransactionUpdater.php  -100
- Gateway/Config/PaymentSdkConfigFactory.php  -100
- Gateway/Service/TransactionServiceFactory.php  -100
- Controller/Frontend/Redirect.php  -96
- Model/Adminhtml/Source/PaymentAction.php  -100
- Gateway/Command/WirecardCommand.php  -87
- Gateway/Request/SepaCreditTransactionFactory.php  -100
- Gateway/Response/ResponseHandler.php  -100
- Gateway/Command/WirecardRefundCommand.php  -96
- Gateway/Validator/AddressAdapterInterfaceValidator.php  -100
- Gateway/Request/AddressFactory.php  -88
- Gateway/Request/PayPalTransactionFactory.php  -97
- Controller/Frontend/Sepamandate.php  -100
- Block/Checkout/PoiPiaBlock.php  -100
- Gateway/Command/WirecardCaptureCommand.php  -96
- Controller/Frontend/Cancel.php  -100
- Gateway/Config/CanCaptureHandler.php  -100
- Gateway/Request/RatepayInvoiceTransactionFactory.php  -100
- Gateway/Request/ItemFactory.php  -100
- Gateway/Model/RetrieveTransaction.php  -100
- Gateway/Validator/QuoteAddressValidator.php  -100
- Model/Adminhtml/Source/PoiPiaAction.php  -100
- Block/Customer/CardRenderer.php  -100
- Gateway/Request/BasketFactory.php  -85
- Gateway/Command/WirecardVoidCommand.php  -96
- Gateway/Request/AlipayXBorderTransactionFactory.php  -100
- Model/Ui/ConfigProvider.php  -100
- Gateway/Request/SofortTransactionFactory.php  -100
- Gateway/Helper/Payment.php  -100
- Gateway/Request/SepaTransactionFactory.php  -100
- Gateway/Request/IdealTransactionFactory.php  -100
         

Complexity increasing per file
==============================
- view/frontend/web/js/view/payment/method-renderer/constants.js  1
- Gateway/Helper/AccountHolderMapper.php  4
- view/frontend/web/js/view/payment/seamless-vault-enabler.js  1
- Block/Checkout/CreditCardThreeDForm.php  1
- view/frontend/web/js/view/payment/method-renderer/seamlessformutils.js  4
- view/frontend/web/js/view/payment/method-renderer/creditcard.js  2
         

Complexity decreasing per file
==============================
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -1
         

Clones removed
==============
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -6
+ view/frontend/web/js/view/payment/method-renderer/creditcard.js  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175&bid=17329626)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
           

Coverage increased per file
===========================
+ Gateway/Request/BasketFactory.php  1
         

Coverage decreased per file
===========================
- Controller/Frontend/Creditcard.php  -2
         

Complexity increasing per file
==============================
- view/frontend/web/js/view/payment/seamless-vault-enabler.js  1
- view/frontend/web/js/view/payment/method-renderer/constants.js  1
- Gateway/Helper/AccountHolderMapper.php  4
- Block/Checkout/CreditCardThreeDForm.php  1
- view/frontend/web/js/view/payment/method-renderer/seamlessformutils.js  4
- view/frontend/web/js/view/payment/method-renderer/creditcard.js  2
         

Complexity decreasing per file
==============================
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -1
         

Clones removed
==============
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -6
+ view/frontend/web/js/view/payment/method-renderer/creditcard.js  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175&bid=17329626)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
           

Coverage increased per file
===========================
+ Gateway/Request/BasketFactory.php  1
         

Coverage decreased per file
===========================
- Controller/Frontend/Creditcard.php  -2
         

Complexity increasing per file
==============================
- Gateway/Helper/AccountHolderMapper.php  4
- view/frontend/web/js/view/payment/method-renderer/creditcard.js  2
- Block/Checkout/CreditCardThreeDForm.php  1
- view/frontend/web/js/view/payment/seamless-vault-enabler.js  1
- view/frontend/web/js/view/payment/method-renderer/seamlessformutils.js  4
- view/frontend/web/js/view/payment/method-renderer/constants.js  1
         

Complexity decreasing per file
==============================
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -1
         

Clones removed
==============
+ view/frontend/web/js/view/payment/method-renderer/creditcard.js  -6
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175&bid=17329626)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
           

Coverage increased per file
===========================
+ Gateway/Request/BasketFactory.php  1
         

Coverage decreased per file
===========================
- Controller/Frontend/Creditcard.php  -2
         

Complexity increasing per file
==============================
- Block/Checkout/CreditCardThreeDForm.php  1
- view/frontend/web/js/view/payment/method-renderer/constants.js  1
- view/frontend/web/js/view/payment/seamless-vault-enabler.js  1
- Gateway/Helper/AccountHolderMapper.php  4
- view/frontend/web/js/view/payment/method-renderer/seamlessformutils.js  4
- view/frontend/web/js/view/payment/method-renderer/creditcard.js  2
         

Complexity decreasing per file
==============================
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -1
         

Clones removed
==============
+ view/frontend/web/js/view/payment/method-renderer/creditcard.js  -6
+ view/frontend/web/js/view/payment/method-renderer/vault.js  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5294175&bid=17329626)

**Pull request merged**